File: app/src/main/java/kztproject/jp/splacounter/viewmodel/AuthViewModel.java
package kztproject.jp.splacounter.viewmodel;

import android.databinding.ObservableField;
import android.support.annotation.StringRes;

import javax.inject.Inject;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;
import kztproject.jp.splacounter.repository.AuthRepository;
import kztproject.jp.splacounter.R;

public class AuthViewModel {

    private final AuthRepository authRepository;
    Callback callback;

    public ObservableField<String> inputString = new ObservableField<>();

    @Inject
    public AuthViewModel(AuthRepository authRepository) {
        this.authRepository = authRepository;
        inputString.set("");
    }

    public void setCallback(Callback callback) {
        this.callback = callback;
    }

    public void login() {
        if (inputString.get().length() == 0 ) {
            callback.showError(R.string.error_login_text_empty);
        } else {
            authRepository.login(inputString.get())
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .doOnSubscribe(disposable -> callback.showProgressDialog())
                    .doOnTerminate(() -> callback.dismissProgressDialog())
                    .subscribe(
                            () -> callback.loginSuccessed(),
                            e -> callback.loginFailed(e));
        }
    }

    public interface Callback {
        void showProgressDialog();

        void dismissProgressDialog();

        void loginSuccessed();

        void loginFailed(Throwable e);

        void showError(@StringRes int stringId);
    }
}

File: app/src/main/java/kztproject/jp/splacounter/viewmodel/AuthViewModel.kt
package kztproject.jp.splacounter.viewmodel

import android.databinding.ObservableField
import android.support.annotation.StringRes
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import kztproject.jp.splacounter.R
import kztproject.jp.splacounter.repository.AuthRepository
import javax.inject.Inject

class AuthViewModel @Inject
constructor(private val authRepository: AuthRepository) {
    private lateinit var callback: Callback

    var inputString = ObservableField<String>()

    init {
        inputString.set("")
    }

    fun setCallback(callback: Callback) {
        this.callback = callback
    }

    fun login() {
        if (inputString.get()!!.isEmpty()) {
            callback.showError(R.string.error_login_text_empty)
        } else {
            authRepository.login(inputString.get()!!)
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .doOnSubscribe { callback.showProgressDialog() }
                    .doOnTerminate { callback.dismissProgressDialog() }
                    .subscribe(
                            { callback.loginSucceeded() }
                    ) { e -> callback.loginFailed(e) }
        }
    }

    interface Callback {
        fun showProgressDialog()

        fun dismissProgressDialog()

        fun loginSucceeded()

        fun loginFailed(e: Throwable)

        fun showError(@StringRes stringId: Int)
    }
}
